### PR TITLE
Use schedule for ongoing, upcoming, and ended comp detection

### LINKED ISSUE #1716: "This competition is ongoing" the day before the competition
I just noticed a small thing when looking at Vancouver Open Summer 2017 competition page.

I get a message saying "Note: This competition is ongoing. Come back after June 10, 2017 to see the results!".

As it is currently 10:05 pm PDT (UTC-8) on June 9th, 2017, I did not expect to see that message :)

### PR DESCRIPTION
**Pull request by UnsolvedCypher**
This PR closes #1716. I'm not exactly sure how to best test this though, so feedback would be appreciated.

### CODE REVIEW

File: WcaOnRails/app/models/competition.rb [hunk 1]
```diff
@@ -946,15 +935,18 @@ def uses_cumulative_across_rounds?
   end
 
   # The name `is_probably_over` is meant to be surprising.
-  # We don't actually know when competitions are over, because we don't know their schedules, nor
-  # do we know their timezones.
-  # See discussion here: https://github.com/thewca/worldcubeassociation.org/pull/1206/files#r98485399.
+  # While we now know competition schedules, not all competitions might have a schedule
   def is_probably_over?
```
Comment (AlbertoPdRF) [inline WcaOnRails/app/models/competition.rb]: Would it make sense to change this method name to `over?` now? Considering that only competitions in the past do not have a schedule, I think that would be reasonable.

File: WcaOnRails/app/models/competition.rb [hunk 2]
```diff
@@ -946,15 +935,18 @@ def uses_cumulative_across_rounds?
...
-    !end_date.nil? && end_date < Date.today
+    last_event_end = top_level_activities.map { |a| a.end_time }.max
+    (!last_event_end.nil? && last_event_end < Time.now) ||
+      (last_event_end.nil? && !end_date.nil? && end_date < Date.today)
   end
 
   def upcoming?
-    !results_posted? && (start_date.nil? || start_date > Date.today)
+    first_event_start = top_level_activities.map { |a| a.end_time }.min
```
Comment (AlbertoPdRF) [inline WcaOnRails/app/models/competition.rb]: This should be `a.start_time`.

File: WcaOnRails/app/models/competition.rb [hunk 3]
```diff
@@ -340,17 +340,6 @@ def warnings_for(user)
     warnings
   end
 
-  def championship_warnings
```
Comment (jfly) [inline WcaOnRails/app/models/competition.rb]: Is there a reason why you removed this?

File: WcaOnRails/app/models/competition.rb (hunk 1 above)
Comment (jfly) [reply WcaOnRails/app/models/competition.rb]: Even if we know the schedule, it might make sense to still use the word "probably", because competitions can run behind schedule =)

(Regardless of what we decide to name this method, we should definitely update the comment about the method to say that.)

File: WcaOnRails/app/models/competition.rb [hunk 4]
```diff
@@ -1020,7 +1012,7 @@ def events_with_round_types_with_results
     ActiveRecord::Base.connection
                       .execute(relation.to_sql)
                       .each(as: :hash).map { |r|
-                        LightResult.new(r)
+                        LightResult.new(r, Country.c_find(r["countryId"]), Format.c_find(r["formatId"]), RoundType.c_find(r["roundTypeId"]), Event.c_find(r["eventId"]))
```
Comment (jfly) [inline WcaOnRails/app/models/competition.rb]: Why did you change this line?

File: WcaOnRails/app/models/competition.rb [hunk 5]
```diff
@@ -946,15 +935,18 @@ def uses_cumulative_across_rounds?
   end
 
   # The name `is_probably_over` is meant to be surprising.
-  # We don't actually know when competitions are over, because we don't know their schedules, nor
-  # do we know their timezones.
-  # See discussion here: https://github.com/thewca/worldcubeassociation.org/pull/1206/files#r98485399.
+  # While we now know competition schedules, not all competitions might have a schedule
   def is_probably_over?
-    !end_date.nil? && end_date < Date.today
+    last_event_end = top_level_activities.map { |a| a.end_time }.max
```
Comment (jfly) [inline WcaOnRails/app/models/competition.rb]: nit: `.map { |a| a.end_time }.max` -> `.map(&:end_time).max`

Comment (jfly) [inline WcaOnRails/app/models/competition.rb]: Do you know what sort of an effect this has on the performance of the `/competitions` page? Is this going to be an `n+1` query?

File: WcaOnRails/app/models/competition.rb (hunk 2 above)
Comment (jfly) [reply WcaOnRails/app/models/competition.rb]: Similar nit as above about using `map(&:start_time)`

File: WcaOnRails/app/models/competition.rb (hunk 1 above)
Comment (AlbertoPdRF) [reply WcaOnRails/app/models/competition.rb]: Yeah, but even in the worst case it can only be a few hours off :stuck_out_tongue:

**Comment by jonatanklosko** [comment]
I'm going to close this due to inactivity.
